### ISSUE TITLE
Implementation of Time and Frequency LSTM

### ISSUE THREAD

**Issue by funguy74** [Microsoft/CNTK#738]
TF LSTM is based on the following Microsoft paper:
https://www.microsoft.com/en-us/research/publication/lstm-time-and-frequency-recurrence-for-automatic-speech-recognition/

From paper, the author said TF-LSTM is implemented in CNTK. I'm wondering how it can be implemented in CNTK.

Basically, F-LSTM and T-LSTM has different operational time scale:

1) For each time t, F-LSTM should swipe log-filter frequency chunks.
2) After finishing frequency swiping, L-LSTM can start operation at time t

So, basically, L-LSTM should wait until collecting all F-LSTM output at time t. How can we implement this in ndl file or brainscript?  What I understand in ndl format is that we can only designate connection and operational units but we cannot assign operational time scale.

Should we modify source code?

**Comment by frankseide** [comment]
To recur over frequency means to iterate over a static dimension. That can currently not be done using CNTK sequence machinery. But you can do it using a combination of BrainScript arrays and Slice().

Here is a piece of code that does something similar. This is from my original BrainScript design sketch, so it is probably not functional without minor changes, but it gives you the idea. This has a recurrence over a fixed number of speech frames that have been stacked into a long vector. Your situation is similar except for the stacking direction:

    // hidden layers
    // Hidden state arrays for all frames are stored in a array object.
    layers[layer:1..numHiddenLayers] = [     // each layer stores a dictionary that stores its output hidden fwd and bwd state vectors
        // model parameters
        W_fwd = Parameter(hiddenDim, featDim)                                           // Parameter(outdim, indim) --in_fwd.rows is an initialization parameter read from the dict
        W_bwd = if layer > 1 then Parameter(hiddenDim, hiddenDim) else Fail("no W_bwd") // W denotes input-to-hidden connections
        H_fwd = Parameter(hiddenDim, hiddenDim)                                         // H denotes hidden-to-hidden lateral connections
        H_bwd = Parameter(hiddenDim, hiddenDim)
        b     = Parameter(hiddenDim, 1)                                                 // bias
        // shared part of activations (input connections and bias)
        z_shared[t:0..T-1] = if layers > 1 then W_fwd * layers[layer-1].h_fwd[t] + W_bwd * layers[layer-1].h_bwd[t] + b    // intermediate layer gets fed fwd and bwd hidden state
                                                      else W_fwd * subframes                                                   + b   // input layer reads frames directly
        // recurrent part and non-linearity
        neededT = if layer < numHiddenLayers then T else centerT+1                  // last hidden layer does not require all frames
        step(H,h,dt,t) = Sigmoid (if (t+dt > 0 && t+dt < T) then z_shared[t] + H * h[t+dt]
                                                           else z_shared[t])
        h_fwd[t:0..neededT-1] = step(H_fwd, h_fwd, -1, t)
        h_bwd[t:T-neededT..T-1] = step(H_bwd, h_bwd,  1, t)
    ]

**Comment by jinyli** [comment]
You needn't modify source code. Brainscript was not avaible when I developed the TF-LSTM which was based on ndl and was too long to describe the recursion along frequency domain. It is better to use Brainscript now. Just modify the Frank's code to do the recursion along frequency domain.  

Note that we recently noticed that current CNTK broke the TF-LSTM training. We are now trying to fix it.

**Comment by funguy74** [comment]
Therefore, by recursively calling LSTM inside layer, we can implement fixed-dimensional F-LSTM in CNTK. It looks like we can easily extend it into TF-LSTM (not separate but simultaneous) by using similar trick.

I really thank both of you. Also, Jinyli, could you notify bug fixed in this thread if it is fixed? I like to try TF-LSTM training when you fix it.

**Comment by jinyli** [comment]
Yes, your understanding is right. We will notify you once the bug is fixed.

**Comment by zhouwangzw** [comment]
CNTK 2.0 provides also Python API which gives you more flexibility to control the training. If you still run into problems, please open a new issue.

**Comment by kaituoxu** [comment]
hi, @funguy74
have you solved this problem?
